### PR TITLE
RFC: introduce runtime representation of broadcast fusion

### LINKED ISSUE #21094: `generated function body is not pure` when using dot notation
I have a generated function in which I rely on Julia's new dot notation for "vectorised" operations on arrays. This results in an error shown below. The minimal working example is

````julia
versioninfo()

@generated function foo(out, x)
    quote
        out .= x .+ x
    end
end

foo([0.0], [1.0])
````

The output is
````
Julia Version 0.6.0-pre.alpha.146
Commit c399558* (2017-03-15 15:20 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin15.6.0)
  CPU: Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
ERROR: LoadError: generated function body is not pure. this likely means it contains a closure or comprehension.
Stacktrace:
 [1] include_from_node1(::String) at ./loading.jl:539
 [2] include(::String) at ./sysimg.jl:14
 [3] process_options(::Base.JLOptions) at ./client.jl:305
 [4] _start() at ./client.jl:371
while loading /Users/davide/Codes/IMEXRK.jl/src/deleteme.jl, in expression starting on line 10
[Finished in 2.1s with exit code 1]
[cmd: ['/Users/davide/Software/julia-0.6/usr/bin/julia', '/Users/davide/Codes/IMEXRK.jl/src/deleteme.jl']]
[dir: /Users/davide/Codes/IMEXRK.jl/src]
[path: /usr/bin:/bin:/usr/sbin:/sbin]
````

The only reference I could find similar to what I get is #19299. This issue seems different to me tough.

### LINKED ISSUE #22053: x.*x not customizable
There doesn't seem to be a way to get `x.*x` to do something different than `x*x`, as the following perhaps naive example shows:

```julia
import Base: *

type T
    t
end

Base.broadcast(::typeof(*), t1::T, t2::T) = 1
*(t1::T, t2::T) = 2
x = T(1)
x*x # Prints 2

x.*x # Also prints 2
```

If there really isn't a way, this is a non-trivial breaking regression in functionality compared to .5 that some packages depended upon.

### LINKED ISSUE #22060: better internal interface for extending `broadcast`
@MikeInnes  suggested in MikeInnes/Flux.jl#31 that we change the dot-call lowering in such a way as to allow it to be disabled for custom types (which may reside on a GPU or something and support only a small set of functions).

For example, currently `x .+ y .* x.^z` is lowered to `broadcast((x,y,z) -> x+y*x^z, x,y,z)`.   This could instead be lowered to:
```jl
if Base.isfusing(x,y,z)
    broadcast((x,y) -> x+y*x^z, x,y,z)
else
    broadcast(+, x, broadcast(*, y, broadcast(^, x, z)))
end
```
with `Base.isfusing(...) = true` being the default.  This would also address #22053 (cc @malmaud).

Pro:

* Makes life easier in the short run for containers based on external libraries that only support a few efficient operations.   They could overload `isfusing` and `broadcast(::typeof(foo), ...)` for a small number of supported `foo`.

* Makes life easier in specialized applications where it may not be *possible* to define `broadcast` for general functions, e.g. [in Convex.jl](https://github.com/JuliaOpt/Convex.jl/issues/192) where it needs to guarantee convexity.

Con:

* You can no longer look at an expression like `x .+ y .* x` and know that it fuses into a single loop with no temporaries.

* There is no middle ground. A single non-fusable operand will "spoil" fusion for an entire set of nested dot calls.  (To fuse a subset of an expression, you'd have to explicitly assign it to a temporary array.)  (We could lower to a nested set of `isfusing` calls, of course, but then you'd get an exponential explosion in lowered code size.)

In the long run, I really think that packages like TensorFlow.jl should exploit the [underlying library's ability to define custom operations as callback functions](https://www.tensorflow.org/extend/adding_an_op) to implement `broadcast(f::Function, ...)` for arbitrary `f`, at which point they can defining `isfusing(...) = true` and get all the benefits of fusion.

### PR DESCRIPTION
**Pull request by vtjnash**
This preserves all dot-fusion information in the type-heirarchy, permitting complex runtime-analysis and re-dispatch overrides. Since no information is obscured in a new type, experiments with variations on #22060 and #22053 can now be implemented in pure Julia just by adding new dispatch rules. Also, since no new types are created, these can now be easily serialized, and can be returned from generated functions.

fix #21094
fix #22060
fix #22053
replaces #22063

### CODE REVIEW

**Comment by vchuravy** [comment]
Awesome! Would be great if it had some tests and examples for actually modifying broadcast behaviour for custom container types.

**Comment by vtjnash** [comment]
Haha. From my description above ("experiments with variations ... can now be implemented"), I was implying that someone else should do that 😛

**Comment by andyferris** [comment]
Cool. This is awesome that this is looked at. :) Thanks @vtjnash 

Since I'm not sure of any preceding discussions - was this approach vs others considered? For example, one might consider lazy but non-parser-fusing broadcast? This problem seems at first blush  possible to deal with using Julia types, dispatch, and dead-simple dot parsing where each dot-call or dot-op makes exactly one `broadcast` call, and broadcasting with arrays could make a `BroadcastedArray` lazily evaluation ed view. The reason I bring this up is because I'm a fan of minimalising parser magic (but then again I'm better at Julia meta programming than at lisp so maybe that's partly me?).

**Comment by stevengj** [comment]
@andyferris, see e.g. #19198 for discussion of making broadcast lazy vs materializing.

File: test/numbers.jl [hunk 1]
```diff
@@ -2914,7 +2914,7 @@ Base.literal_pow(::typeof(^), ::PR20530, ::Val{p}) where {p} = 2
     p = 2
     @test x^p == 1
     @test x^2 == 2
-    @test [x,x,x].^2 == [2,2,2]
+    @test_broken [x, x, x].^2 == [2, 2, 2] # literal_pow violates referential transparency
```
Comment (stevengj) [inline test/numbers.jl]: What result does this give now?

**Comment by timholy** [comment]
Probably also the best path for fixing https://github.com/JuliaLang/julia/pull/22932#issuecomment-330711997

File: test/numbers.jl (hunk 1 above)
Comment (StefanKarpinski) [reply test/numbers.jl]: On master, I get `[x,x,x].^2 == [2,2,2]` but I guess this PR breaks this? Temporarily or permanently?

**Comment by stevengj** [comment]
What happens in this PR to `@. x^2 + 2x + sin(x)`?  Currently, that gets lowered to `broadcast(x -> x^2 + 2x + sin(x), x)`, i.e. the multiple appearances of `x` are collapsed into a single argument.  Is that still true?

File: test/numbers.jl (hunk 1 above)
Comment (TotalVerb) [reply test/numbers.jl]: I don't think `literal_pow` really played well with `broadcast` and dot syntax to begin with. It was a little unpredictable when `literal_pow` or regular `^` would be called...

**Comment by mbauman** [comment]
No, they are currently passed as three arguments:

```
julia> expand(Main, :(@. x^2 + 2x + sin(x)))
:((Base.broadcast)(((Core.apply_type)((Core.getfield)(Base.Broadcast, :Fusion), 3, false))(((Core.getfield)(Base.Broadcast, :FusionCall))(+, (Core.tuple)(((Core.getfield)(Base.Broadcast, :FusionCall))(^, (Core.tuple)(((Core.apply_type)((Core.getfield)(Base.Broadcast, :FusionArg), 1))(), ((Core.getfield)(Base.Broadcast, :FusionConstant))(2))), ((Core.getfield)(Base.Broadcast, :FusionCall))(*, (Core.tuple)(((Core.getfield)(Base.Broadcast, :FusionConstant))(2), ((Core.apply_type)((Core.getfield)(Base.Broadcast, :FusionArg), 2))())), ((Core.getfield)(Base.Broadcast, :FusionCall))(sin, (Core.tuple)(((Core.apply_type)((Core.getfield)(Base.Broadcast, :FusionArg), 3))()))))), x, x, x))
```

Are there cases where that makes a significant performance difference?

File: test/numbers.jl (hunk 1 above)
Comment (mbauman) [reply test/numbers.jl]: Yes, on this PR you get `[1, 1, 1]`.

**Comment by stevengj** [comment]
@mbauman, I thought that it did, but now I'm trying a simple example and I'm not seeing a difference:

```jl
julia> f!(dest, x) = broadcast!(x -> x^2 + 3x - 5x^3, dest, x);
julia> g!(dest, x) = broadcast!((x,y,z) -> x^2 + 3y - 5z^3, dest, x,x,x);
julia> x = rand(10^4); y = similar(x);

julia> @btime f!($y, $x);
  7.167 μs (0 allocations: 0 bytes)

julia> @btime g!($y, $x);
  7.183 μs (0 allocations: 0 bytes)
```

**Comment by stevengj** [comment]
It would also be nice to see a proof-of-concept implementation of a custom array type for which fusion is disabled, just to make sure that this is feasible with a reasonable amount of code given the new design.

**Comment by mbauman** [comment]
That's precisely what I'm working on right now. I understand this much better after playing with it for a little bit.  Here's a simple example:

```
@. 2x + 1
```

Expands out into:

```
julia> d = Base.Broadcast.Fusion{1,false}(
               Base.Broadcast.FusionCall(+,
                   (Base.Broadcast.FusionCall(*,
                       (Base.Broadcast.FusionConstant(2),
                        Base.Broadcast.FusionArg{1}())),
                    Base.Broadcast.FusionConstant(1))))
(a_1) -> +(*(2, a_1), 1)

julia> x = 1:2;

julia> broadcast(d, x)
2-element Array{Int64,1}:
 3
 5
```

So, now I think you'll need to overload `broadcast(::Fusion, ::UnitRange)` and dig through the call graph.  It doesn't look like it'll be easy…

It'd be really nice if we could pass the objects themselves to the `FusionArg` and then eagerly evaluate them in cases like this.  But that won't play nicely with broadcasting.  Still puzzling this through...

**Comment by mbauman** [comment]
Ok, I don't have the complete thread here yet, but it looks like we'll want to add a recursive pre-walk to make it easy to modify the call tree and its arguments.  For example, we want to recursively identify that, e.g., `FusionCall(+, ::FusionConstant{Int}, ::FusionArgument{1})` is a representation of `unfused_broadcast(+, ::Int, ::UnitRange)`.

In this case, we'd then swap the `FusionCall` out in exchange for simply the `FusionArgument{1}()`, and then eagerly evaluate the first fusion argument to `r.start+i:r.stop+i`.

Does this plan of attack sound reasonable and achievable?  I definitely will need more time to process how this becomes an extensible interface.  Others are very welcome to pitch in. :)

**Comment by stevengj** [comment]
@mbauman, if this analysis of the fusion call graph occurs at runtime, is there an issue with type-stability?  Functions that disable fusion still need to be inferrable.

**Comment by mbauman** [comment]
Right, but that's exactly what we'd need to address something like https://github.com/JuliaLang/julia/pull/22932#issuecomment-330711997.  I was simply trying to identify if/how it'd be possible to get there from the current implementation in a way that's easy to extend.

This is definitely an improvement in many cases… but it blows up fast due to all the nested types.  For example, this definition is no longer inferable:

```
julia> f(x) = @. x^5 - 3x^4 + 2x^3 + 10x^2 - 4x + 64

julia> @code_warntype f(1:10)
Variables:
  #self# <optimized out>
  x::UnitRange{Int64}
  T::Type{#s50} where #s50
…
      return #temp#@_53::Any
  end::Any
```

**Comment by timholy** [comment]
I merged this into master. Here's a fun demo.

### code
```julia
function Base.broadcast(f, r::UnitRange)
    if all_ops_ursafe(f)
        UnitRange(f(first(r)), f(last(r)))
    elseif all_ops_rsafe(f)
        range(f(first(r)), f(step(r)), length(r))
    else
        broadcast(f, Broadcast.DefaultArrayStyle{1}(), nothing, nothing, r)
    end
end

all_ops_ursafe(f::Broadcast.Fusion) = all_ops_ursafe(f.f)
all_ops_ursafe(f::Broadcast.FusionCall) = all_ops_ursafe(f.f) && all(all_ops_ursafe, f.args)
all_ops_ursafe(::typeof(+)) = true
all_ops_ursafe(::typeof(-)) = true
all_ops_ursafe(f::Broadcast.FusionConstant{<:Real}) = true
all_ops_ursafe(f::Broadcast.FusionArg) = true
all_ops_ursafe(arg) = false

all_ops_rsafe(f::Broadcast.Fusion) = all_ops_rsafe(f.f)
all_ops_rsafe(f::Broadcast.FusionCall) = all_ops_rsafe(f.f) && all(all_ops_rsafe, f.args)
all_ops_rsafe(::typeof(+)) = true
all_ops_rsafe(::typeof(-)) = true
all_ops_rsafe(::typeof(*)) = true
all_ops_rsafe(::typeof(/)) = true
all_ops_rsafe(f::Broadcast.FusionConstant) = true
all_ops_rsafe(f::Broadcast.FusionArg) = true
all_ops_rsafe(arg) = false
```
### demo
```
julia> using Test

julia> x = 1:3
1:3

julia> @inferred(2 .+ x .+ 1)
4:6

julia> @inferred(2 .* x .+ 1)
3:2:7

julia> @inferred(sin.(x) .+ 1)
3-element Array{Float64,1}:
 1.8414709848078965
 1.9092974268256817
 1.1411200080598671
```

Overall I'm pretty optimistic about this and think we probably want this for 1.0. My two concerns:
- this would need updating for the change in keyword arguments, and I'm a bit skeptical that I have time to tackle that within the next 48 hours (I'm in the middle of a move...)
- the quality of the generated code doesn't seem great, though I didn't have time to check in detail

**Comment by vtjnash** [comment]
I suppose this is OK, but I don't really like this PR anymore, since it seemed like it only addressed part of the problem. I started a new project at https://github.com/JuliaLang/julia/compare/master...vtjnash:jn/lazydotfuse to experiment with being fully lazy and it seemed better. It needs updating for the changes to the broadcast API.

**Comment by Sacha0** [comment]
> Here's a fun demo.

Neat demo `@timholy`! :)

**Comment by timholy** [comment]
I didn't spend a lot of time looking it over, but IIUC `jn/lazydotfuse` seems to generalize the approach of [MappedArrays.jl](https://github.com/JuliaArrays/MappedArrays.jl). As someone whose computing life routinely features 1TB 4d arrays, I *love* lazy operations. But, the problem is, when do you evaluate? Will users get their hands on a `Broadcast` object, or is this something that gets evaled at the "end" of broadcasting? (My Scheme is nonexistent, so I can't easily tell from just looking at the code.) If so, won't the fact that a `Broadcast` is neither `<:AbstractArray` nor `<:Tuple` cause downstream problems?

Conversely, this PR doesn't seem to suffer from that problem at all. What specifically seemed missing to you?

**Comment by vtjnash** [comment]
> Will users get their hands on a Broadcast object, or is this something that gets evaled at the "end" of broadcasting? 

The operations are realized at the "end" of the broadcasting, so the effect of it is transparent to user code.

This PR still does an awkward amount of work in scheme. The other branch eliminates that and moves everything into Julia. The net result is that the default is for Julia to build effectively the same representation as this PR. But allowing libraries to implement more accurate control over the transform.

**Comment by StefanKarpinski** [comment]
So it seems like what @vtjnash has done on that branch is exactly what I had envisioned when I proposed doing this with lazy operations instead of as a purely syntactic transformation. The key insight is that we were already effectively deciding where the materialization boundary is in the syntactic approach: the materialization boundary is exactly where dot fusion stops.

**Comment by timholy** [comment]
FYI I have `vtjnash:jn/lazydotfuse` rebased on recent master and passing `test/broadcast.jl`. Performance kinda stinks right now (4x regression for operations like `x .* (x .+ 1) .+ 2`), and fixing that might take some redesign. I mention it because of its potential relevance for 1.0.

**Comment by vtjnash** [comment]
> that might take some redesign

Can you give a quick idea of what part of it is problematic? I had done some work to make sure the performance tests such as `sparse/higherorderfns` were passing on the original commit.

**Comment by timholy** [comment]
It's (at least) the calls to `newindexer` in `_broadcast_getindex(bc::Broadcasted, I)`. That only gets used if you have "nested" broadcasting, so it's possible that wasn't triggered in the tests you ran? Either that, or the rebase is hurting something that you had working nicely. (It's not just a simple rebase, since the API changed, so it's entirely possible I messed something up.)

There's an obvious fix, to cache the result of `newindexer` in a `Broadcasted`. That way it runs once upon construction rather than with each element access.

But I suspect there might be some further rework that is needed. For example, https://github.com/JuliaArrays/StaticArrays.jl/pull/348#issuecomment-350101113 demonstrated that computing `shape` hurts performance substantially in some cases. The new broadcasting API conveniently made it easy to skip that operation, but it's less clear how to do that with `Broadcasted`, especially because the obvious answer (omitting `shape` as a field) is in the exact opposite direction of caching the result of `newindexer`. It may take a little bit of thinking about how to get everything we want in all cases and still keeping a nicely extensible API.

I'm reasonably confident that I'll get there, though. The two risks are (1) the holiday season will cut into development time and (2) it seems likely we're going to end up with an API that will be substantially changed from the new one [documented here](https://docs.julialang.org/en/latest/manual/interfaces/#man-interfaces-broadcasting-1). The good news is that the hardest-to-design yet best part of the API---the creation of `BroadcastStyle` and the splitting out of `broadcast_similar`---seems likely to be something we can largely retain.

**Comment by timholy** [comment]
I'm going to post my current thoughts about how this will change the API. Right now I'm in "artistic" development mode, inspired by the fact that I grabbed the wrong laptop charger on my way out of town for the holidays and thus can't test actual code. (I'm using a graciously-loaned laptop that can't yet build julia from source, though it may be close.) Before leaving I was pretty close to having something functional, extensible, and decently performant, so I think it's appropriate to show where we are and discuss the design even in the abstract.

Here's a [gist](https://gist.github.com/timholy/11dfd23efdd2e8d51ed47b2c93194d11) linking to some mostly-working code. As in Jameson's original work, a `Broadcasted` is a lazy wrapper for a broadcasting operation, and a `TupleLL` (renamed from Jameson's `TypeTuple`) is a linked-list representation of a tuple (like in lisp) that unlike regular tuples is also inferable for elements that are types. (This is necessary for operations like `ceil.(Int, x)` where the `Int` arg must be handled by broadcasting machinery.)

The main changes to the API should be the following:

- simple cases: users define a `BroadcastStyle` and define `broadcast_similar`, as before. But we'll have to ditch the `f` argument from broadcast_similar, because nesting `Broadcasted` objects means there is not necessarily a single `f` anymore. For example, in `x.*(x .+ 1)` the `x .+ 1` will be nested inside the `x.*y`, and there is no fused `f`.
- likewise, after the style, eltype, and indices/axes arguments to `broadcast_similar`, we'll probably just pass the `Broadcasted` object itself. In cases where `broadcast_similar` has to reference one of the input args (see [the `ArrayAndChar` example](https://docs.julialang.org/en/latest/manual/interfaces/#man-interfaces-broadcasting-1)), most likely users are going to have to learn how to walk a `TupleLL` (we could unpack in a tuple but that loses information about how it's nested with respect to `f`, and for operations involving types it will make things non-inferrable).
- to solve performance problems like the one I mentioned above for StaticArrays.jl, I introduced an `instantiate` cascade that package authors should be able to short-circuit to avoid unnecessary computations. As currently designed, it will start from whatever information is present and incrementally fill in the missing pieces (repeatedly re-wrapping the data as each new piece gets added), but one can use dispatch to interrupt the cascade at any point.
- `broadcast` and `broadcast!` become almost trivial, and to me it seems that once the lazy-wrapper is built the part that does the actual work should be called `copy` and `copyto!`. When needed, users will specialize `copy(::Broadcasted{Style})` instead of `broadcast`, and `copyto!(dest, ::Broadcasted{Style})` instead of `broadcast!`

Potential concerns:
- broadcasting works for tuples as well as arrays. Currently we don't define `copy(::Tuple)`. Is `copy(::Broadcasted{Style{Tuple}})::Tuple` acceptable, or an abuse?
- walking a `TupleLL` (and diving into a nested `Broadcasted` argument) is more complex than processing a list of args.
- performance. There just seem to be weird little "gotchas" (see my comment about `@noinline` in the gist) that can hopefully be ironed out. I still haven't succeeded in recovering full performance for a trivial operation like a 2x2 .+ scalar, though I was close (14ns vs 18ns, though that was for an `instantiate`d objected so it was a bit of a cheat). I'm worried about the heap-vs-stack thing for creating wrappers, which this currently does with abandon, and whether the escape analysis will be sufficient to elide the wrappers. (For the ambiguity-reducing strategy of creating a `Broadcasted{Nothing}` for those cases where users only want to specialize `copyto!` on `typeof(dest)`, the cost of creating the type seemed to be something that wasn't elided.) We can fix these by uglifying the API (e.g., `copyto!(dest, bc, nothing)` rather than re-wrapping `bc`), but I'm not very happy about that. I would particularly welcome thoughts about this aspect of the API.

**Comment by bramtayl** [comment]
I wonder if it would be too breaking to make broadcasting fully lazy. These `copyto` functions look suspiciously like `collect`s.

**Comment by bramtayl** [comment]
PS can't we just make regular tuples inferable for types?

**Comment by timholy** [comment]
One of the interesting features of this approach is that you can make it lazy, but currently it requires that you manually construct the `Broadcasted` object. That's not very hard, but perhaps we could even use `@views bc = ...` as a way of getting the lazy object without allocating substantial new storage.

And yes, these `copy`/`copyto!` calls are indeed acting essentially as `collect`s. One small difference is that `collect` explicitly "casts" to 1-based indexing, whereas the `copy`/`copyto!` preserves the indices.

**Comment by timholy** [comment]
> PS is it possible to make regular tuples inferable for types?

I think that was tried earlier (I believe this changed at around 0.4 or perhaps 0.5), but for reasons I don't clearly recall it led to problems. This is the reason that we have tuples and tuple-types, i.e., `typeof((1, 1.0)) == Tuple{Int64,Float64}` rather than `(Int, Float64)`.

**Comment by bramtayl** [comment]
Cf https://github.com/JuliaLang/julia/issues/16029#issuecomment-353517423

**Comment by Sacha0** [comment]
Similarly https://github.com/JuliaLang/julia/pull/25217#issuecomment-353438898 :).

**Comment by bramtayl** [comment]
It might be worth exporting the linked lists as a standard library package?

**Comment by KristofferC** [comment]
It seems the questions here are diverging further and further from the core of this PR. Please open new, focused, issues instead and cross reference back to this one. Issues are cheap.

**Comment by timholy** [comment]
What do folks think I should do about this? I'm currently blocked on #25267. I could push a branch that has the requisite changes in it, but it is going to fail nanosoldier. Or if it doesn't, it's only because we don't have sufficiently comprehensive broadcasting tests.

I guess the real question is whether we want to merge something that has known performance regressions for the sake of the API change, and assume that the performance fix will come.

**Comment by vchuravy** [comment]
I would say do the API changes now and performance improvements during the alpha.

**Comment by stevengj** [comment]
Can't this be done for 1.1?  Right now, there is no good API for defining your own `broadcast` containers other than overloading `broadcast` itself.   So, adding an API for this would not be breaking.  I'd rather not merge something at the last minute that massively slows things down for an API that we've barely tested.

**Comment by StefanKarpinski** [comment]
I agree, @stevengj – hooking into broadcast should be explicitly not a stable API in 1.0.

**Comment by timholy** [comment]
Agreed this is a tough call. You outline the dangers well. The flip side is that it's a bit unsatisfying to say "Julia 1.0 is here! It's got a stable API except for broadcasting!" I suspect the coming PR is the last remaining component needed to avoid having that kind of caveat, and to me that's sufficiently attractive that I'm going to keep plugging away at this, get the PR submitted, and y'all can decide what to do with it.

If the goal were only to allow broadcasting to be lazy, then I'd say that's not important enough to worry about. But as a refresher, the real problems to be fixed are:
- it's pretty awful that now we return an `Array` for `(1:5) .+ 1`. We could put a bandaid on that by specializing `broadcast(::typeof(+), ::UnitRange, ::Real)`, but due to the lisp-fusion things like `2.*(1:5) .+ 1` are beyond repair, and even the simple fix would behave differently for variables than for literals.
- the lisp-fusion makes life essentially impossible for, say, the GPU folks or anyone else who calls a C library, who can't make use of a fused operation and can't easily unfuse it.

If we do this later, then my crystal ball suggests it might look something like this:
- we'll have to add a `broadcast_similar` method that throws an error if the API documented [here](https://docs.julialang.org/en/latest/manual/interfaces/#man-interfaces-broadcasting-1) is used---without the lisp-fusion it's not clear what to supply for `f`, so I don't think we can issue a traditional deprecation warning (in the sense that code will continue to function).
- There seems also to be concern about the impact on `literal_pow` [above](https://github.com/JuliaLang/julia/pull/23692#pullrequestreview-63969525), though I'm not far enough along to have anything useful to say about that.
-  The return type of many broadcasting operations that involve ranges will change. To me that's a breaking change that goes beyond "the API for hooking into broadcasting has changed."

If none of these sound serious to others, don't let this stand in the way of the 0.7-alpha release.

**Comment by stevengj** [comment]
@timholy, the minimal change to address some of those issues would be something like #22063: a way to disable fusion for certain argument types, without any more complicated stuff like laziness or this PR.

If we ever change broadcast to be fully lazy or whatever, we can still implement `isfusing` on top of that so that container types still have an easy way to disable fusion.

(Honestly, I don't see the big deal about ranges.  Just don't use `.*` if you want a range result; use `*` instead.)

**Comment by StefanKarpinski** [comment]
> The flip side is that it's a bit unsatisfying to say "Julia 1.0 is here! It's got a stable API except for broadcasting!"

It would certainly be better to be able to say "everything is perfect and we won't change it", but we're really out of time to keep working on this and it doesn't seem ready yet. Since broadcasting and dot syntax are such new and novel features, I don't think it's the end of the world to say that:

* Usage of broadcasting will continue to work from user's perspective the way it does now;
* Implementation of broadcasting may change  and that may break libraries that hook into it deeply.

For the vast majority of Julia users, the "broadcasting API" is how broadcasting behaves and how they use it, not how it's implemented and extended. If/when we do change the internals of broadcasting, we can work closely with packages that hook into it to ensure a smooth transition so that non-library-author Julia users are none the wiser that anything has changed.

**Comment by vtjnash** [comment]
> There seems also to be concern about the impact on literal_pow above, though I'm not far enough along to have anything useful to say about that.

It's possible to handle, it just requires some code duplication.

**Comment by andreasnoack** [comment]
Do we still need all the `literal_pow` stuff now that we have constant propagation?

**Comment by StefanKarpinski** [comment]
@andreasnoack, please don't derail this conversation to rehash that. I've explained why that's needed elsewhere and will be happy to do so again but not here.

**Comment by timholy** [comment]
New version of this PR in #25377.

@stevengj 
> (Honestly, I don't see the big deal about ranges. Just don't use .* if you want a range result; use * instead.)

We're deprecating `a + 1` for other `AbstractVector`s, so telling people to use `*` for ranges and `.*` for all other `AbstractVector`s is not a satisfying solution. We seem to have missed deprecating those methods for ranges, but I do that as a separate commit in #25377.

That fact then makes the alternative less attractive:
> the minimal change to address some of those issues would be something like #22063: a way to disable fusion for certain argument types, without any more complicated stuff like laziness or this PR.

So we'd have to disable fusion for all ranges, and that presumably means ending fusion for quite a few objects. Totally doable, of course, just unattractive.

**Comment by stevengj** [comment]
@timholy, as you mentioned above, we could define `broadcast(::typeof(+), ::Range, ::Real)` so that people could get a range from `(1:5) .+ 1` etcetera.    If they combine this with multiplication, they just need to use `*`, e.g. `2(1:5) .+ 1` and not `2 .* (1:5) .+ 1` if they want a range result.  Anything more complicated with a range would produce an an array.  This doesn't seem so bad to me.

**Comment by timholy** [comment]
One problem with `broadcast(::typeof(+), ::Range, ::Real)` is it doesn't work for `r::Range .+ n::Number` when `n` is literal, because literals get fused into f. On Julia 0.6:
```julia
julia> Base.broadcast(::typeof(+), ::UnitRange, ::Int) = "crazy"

julia> (1:5) .+ 1
5-element Array{Int64,1}:
 2
 3
 4
 5
 6

julia> n = 1
1

julia> (1:5) .+ n
"crazy"
```

But I agree that deep nesting is rare, and I could live without preserving the range in such cases.

**Comment by stevengj** [comment]
Okay, then let's just define `+ number` for ranges.   Ranges are a special type unto themselves, not a linear-algebra object, so I don't see a problem with defining special operations on them.

**Comment by timholy** [comment]
We have that currently, so that wouldn't be a change. But that does take us back to

> We're deprecating a + 1 for other AbstractVectors, so telling people to use + for ranges and .+ for all other AbstractVectors is not a satisfying solution.

That said, not everything in 1.0 is going to be satisfying. Still, this problem and others are fixed in #25377. It's just a question of whether people are scared by it, and of course to finish up the loose ends.

As an alternative, what about removing the literal-slurping from the lisp fusion code? Are there big negatives? That would at least make broadcasting work in simple cases.

**Comment by stevengj** [comment]
> Still, this problem and others are fixed in #25377.

At a huge performance cost, I thought you said?

**Comment by stevengj** [comment]
> As an alternative, what about removing the literal-slurping from the lisp fusion code? Are there big negatives? That would at least make broadcasting work in simple cases.

Last I checked, there was a significant performance penalty to simple expressions if you remove the literal-slurping.  But this doesn't seem to be the case any more?
```
julia> f!(x) = broadcast!(x -> x^2, x, x);
julia> g!(x) = broadcast!((x,p) -> x^p, x, x,2);
julia> x = zeros(10^3);
julia> @btime f!(x); @btime g!(x);
  90.810 ns (0 allocations: 0 bytes)
  96.155 ns (0 allocations: 0 bytes)

julia> f!(x) = broadcast!(x -> 3x^2 + 4x + 5, x, x);
julia> g!(x) = broadcast!((x,c1,c2,c3,c4) -> c1*x^c2 + c3*x + c4,x, x,3,2,4,5);
julia> @btime f!(x); @btime g!(x);
  198.718 ns (0 allocations: 0 bytes)
  199.601 ns (0 allocations: 0 bytes)
```
I'm not sure what changed.

Another problem with removing literal-slurping is that it eliminates the type-stability of Unitful expressions involving literal powers.  One possibility would be to *only* do literal-slurping for literal powers.

**Comment by timholy** [comment]
> At a huge performance cost, I thought you said?

Not as bad now. For `a = rand(2,2)`, master gives
```julia
julia> @btime broadcast!(+, $b, $a, 1);
  12.715 ns (0 allocations: 0 bytes)
```
and #25377 gives
```julia
julia> @btime broadcast!(+, $b, $a, 1);
  17.991 ns (0 allocations: 0 bytes)
```

**Comment by timholy** [comment]
> One possibility would be to only do literal-slurping for literal powers.

If we don't do #25377, that would be an improvement. Then, #22063 might be enough to prevent the GPU people from hating us. You could conceivably rework it around `BroadcastStyle` since that already has `promote_type`-like behavior (see the docs about binary rules and `combine_styles`).

Still less flexible than #25377, but it's a much smaller change, and it automatically fixes the `literal_pow` problem.

File: base/broadcast.jl [hunk 2]
```diff
@@ -859,4 +859,229 @@ macro __dot__(x)
...
+    fargs = tuplehead(args, Val(Nreq))
+    vararg = tupletail(args, Val(Nreq))
+    return f.f((fargs..., vararg)...)
+end
+@inline (f::FusionArg{N})(args...) where {N} = args[N]
+@inline (f::FusionConstant)(args...) = f.c
+@inline (f::FusionCall)(args...) = f.f(map(a -> a(args...), f.args)...)
+# TODO: calling _apply on map _apply is not handled by inference
+# for now, we unroll some cases and generate others, to help it out
+#@inline (f::FusionApply)(args...) = Core._apply(f.f, map(a -> a(args...), f.args)...)
```
Comment (vtjnash) [inline base/broadcast.jl]: equivalently, I think this is also `f.f(flatten(map(a -> a(args...), f.args)...)...)`
where `flatten(args...) = Core._apply(tuple, args...)`

**Pull request closed without merging**